### ISSUE TITLE
Uploading a single measurement with a specified time

### ISSUE THREAD

**Issue by nobodyinperson** [sensebox/openSenseMap-API#169]
Maybe it's me, but I am not able to [upload a single measurement](https://docs.opensensemap.org/#api-Measurements-postNewMeasurement) **with a specified time** via the API.

## Expected Behavior

The API should accept the time/measurement pair.

## Current Behavior

The API fails with `{"code":"InternalServer","message":"Cannot set property 'timestamp' of undefined"}`.

## Possible Solution

?

## Steps to Reproduce (for bugs)

```bash
curl \
    -H "Content-Type: application/json" \
    -d '{"value":"3333","createdAt":"2018-08-28T10:00:00.000Z"}' \
    https://api.opensensemap.org/boxes/$SENSEBOX_ID/$SENSOR_ID
# {"code":"InternalServer","message":"Cannot set property 'timestamp' of undefined"}% 
curl -V
# curl 7.47.0 (x86_64-pc-linux-gnu) libcurl/7.47.0 GnuTLS/3.4.10 zlib/1.2.8 libidn/1.32 librtmp/2.3
# Protocols: dict file ftp ftps gopher http https imap imaps ldap ldaps pop3 pop3s rtmp rtsp smb smbs smtp smtps telnet tftp 
# Features: AsynchDNS IDN IPv6 Largefile GSS-API Kerberos SPNEGO NTLM NTLM_WB SSL libz TLS-SRP UnixSockets
```

(Also, there seems to be a bug in `curl 7.58.0` where the error message is just not displayed...)

## Context

I am developing the [Python package `sensemapi`](https://pypi.org/project/sensemapi/) to access the OpenSenseMap API.

## Your Environment

* Link to your project: https://gitlab.com/tue-umphy/python3-sensemapi

**Comment by mpfeil** [comment]
Hi @nobodyinperson 

thanks for opening this issue. I tried to reproduce the error by using your `curl` command. I copied it and just replaced `$SENSEBOX_ID` and `$SENSOR_ID` with actual IDs and the measurement is saved.

I was using `curl 7.54.0`. I am going to investigate tomorrow and let you know if I find something.

**Comment by nobodyinperson** [comment]
Thanks @mpfeil!

I detected that using the current time for `createdAt` works fine:

```bash
curl \         
    -H "Content-Type: application/json" \
    -d '{"value":"3333","createdAt":"'`date -u +%FT%TZ`'"}' \
    https://api.opensensemap.org/boxes/$SENSEBOX_ID/$SENSOR_ID
# "Measurement saved in box"
```

Maybe using dates prior to the box creation breaks something?

**Comment by ubergesundheit** [comment]
Note that `date -u +%FT%TZ` doesn't contain microseconds `.000`. This is certainly a bug. Thanks for reporting

**Comment by mpfeil** [comment]
> Maybe using dates prior to the box creation breaks something?

Thanks for the hint. I am going to check it. When I tested your `curl` command it also worked with the old `createdAt`.

**Comment by nobodyinperson** [comment]
Okay, I did some testing based on my [sensemapi](https://gitlab.com/tue-umphy/python3-sensemapi) package.

**TL;DR**:  Uploading measurements more than 20-30 seconds **prior** to the box creation date seems to be a problem.


# Uploading single past measurements

## Import the sensemapi package


```python
import sensemapi
import datetime
import time
```

## Access an account


```python
account = sensemapi.account.SenseMapAccount( 
    email = os.environ.get("SENSEMAP_EMAIL"),
    password = os.environ.get("SENSEMAP_PASSWORD"),
    api = "https://api.opensensemap.org",
    )
```

## Create a new senseBox


```python
# Create a senseBox
box = sensemapi.senseBox.senseBox(
    name = "Test senseBox",
    exposure = "outdoor",
    current_lat = 53.57498,
    current_lon = 9.98635,
    )
# add a sensor
box.new_sensor(
    title="temperature",
    unit="°C",
    type="SHT31",
    icon="osem-temperature-celsius",
    )
# upload the senseBox
account.new_box(box)
# register the box creation time
box_creation_time = datetime.datetime.now()
box_creation_time
# wait a little...
time.sleep(1)
```

## Upload measurements

### Access the sensor


```python
account.get_own_boxes()
box = account.boxes[0]
sensor = box.sensors.temperature
```

### Upload a measurement at the creation time


```python
sensor.last_time = box_creation_time
sensor.last_value = 10
try:
    sensor.upload_measurement()    
except sensemapi.errors.OpenSenseMapAPIError as e:
    print(e)
sensor.get_measurements()
```




    createdAt
    2018-08-29 14:31:15+00:00    10
    Name: temperature (SHT31) [°C], dtype: int64



### Upload a measurement past the creation time


```python
sensor.last_time = datetime.datetime.now()
sensor.last_value = 1
try:
    sensor.upload_measurement()    
except sensemapi.errors.OpenSenseMapAPIError as e:
    print(e)
sensor.get_measurements()
```




    createdAt
    2018-08-29 14:31:17+00:00     1
    2018-08-29 14:31:15+00:00    10
    Name: temperature (SHT31) [°C], dtype: int64



### Upload a measurement seconds prior to the creation time


```python
for i in range(1,100,10):
    sensor.last_time = box_creation_time - datetime.timedelta(seconds=i)
    print(sensor.last_time)
    sensor.last_value = i
    try:
        sensor.upload_measurement() 
    except sensemapi.errors.OpenSenseMapAPIError as e:
        print("{} seconds before creation: {}".format(i,e))
sensor.get_measurements()
```

    2018-08-29 16:31:14.507246
    2018-08-29 16:31:04.507246
    2018-08-29 16:30:54.507246
    2018-08-29 16:30:44.507246
    31 seconds before creation: Posting measurement didn't work: Cannot set property 'timestamp' of undefined
    2018-08-29 16:30:34.507246
    41 seconds before creation: Posting measurement didn't work: Cannot set property 'timestamp' of undefined
    2018-08-29 16:30:24.507246
    51 seconds before creation: Posting measurement didn't work: Cannot set property 'timestamp' of undefined
    2018-08-29 16:30:14.507246
    61 seconds before creation: Posting measurement didn't work: Cannot set property 'timestamp' of undefined
    2018-08-29 16:30:04.507246
    71 seconds before creation: Posting measurement didn't work: Cannot set property 'timestamp' of undefined
    2018-08-29 16:29:54.507246
    81 seconds before creation: Posting measurement didn't work: Cannot set property 'timestamp' of undefined
    2018-08-29 16:29:44.507246
    91 seconds before creation: Posting measurement didn't work: Cannot set property 'timestamp' of undefined





    createdAt
    2018-08-29 14:31:17+00:00     1
    2018-08-29 14:31:15+00:00    10
    2018-08-29 14:31:14+00:00     1
    2018-08-29 14:31:04+00:00    11
    2018-08-29 14:30:54+00:00    21
    Name: temperature (SHT31) [°C], dtype: int64



## Delete all senseBoxes from the account


```python
account.get_own_boxes()
for box in account.boxes:
    print("deleting box {}".format(box.id))
    account.delete_box(box.id, really = True)
```

    deleting box 5b86ae1a7c519100197e71d5

**Comment by noerw** [comment]
Conditions to reproduce:
- box where `exposure == 'mobile'`
- measurements without `location` field, and with `timestamp` earlier than box creation, no matter when exactly

For mobile boxes, the timestamp of the first location is dated back to the measurement timestamp, when a measurement is submitted with timestamp but without coordinates.
There seems to be an error in that logic.

https://github.com/sensebox/openSenseMap-API/blob/ccf0f8d/packages/models/src/box/box.js#L387-L391
